examples/nlp/intent_detection_slot_tagging/joint_intent_slot_infer.py

```diff
@@ -94,38 +108,72 @@
     concatenate(tensors) for tensors in evaluated_tensors
 ]
 
-pred_intents = np.argmax(intent_logits, 1)
-logging.info('Intent prediction results')
-
-intents = np.asarray(intents)
-pred_intents = np.asarray(pred_intents)
-intent_accuracy = sum(intents == pred_intents) / len(pred_intents)
-logging.info(f'Intent accuracy: {intent_accuracy}')
-logging.info(classification_report(intents, pred_intents))
-
-slot_preds = np.argmax(slot_logits, axis=2)
-slot_preds_list, slot_labels_list = [], []
+# slot accuracies
+logging.info('Slot Prediction Results:')
+slot_preds_unmasked = np.argmax(slot_logits, axis=2)
 subtokens_mask = subtokens_mask > 0.5
-for i, sp in enumerate(slot_preds):
-    slot_preds_list.extend(list(slot_preds[i][subtokens_mask[i]]))
-    slot_labels_list.extend(list(slot_labels[i][subtokens_mask[i]]))
-
-logging.info('Slot prediction results')
-slot_labels_list = np.asarray(slot_labels_list)
-slot_preds_list = np.asarray(slot_preds_list)
-slot_accuracy = sum(slot_labels_list == slot_preds_list) / len(slot_labels_list)
-logging.info(f'Slot accuracy: {slot_accuracy}')
-logging.info(classification_report(slot_labels_list, slot_preds_list))
+slot_labels = slot_labels_unmasked[subtokens_mask]
+slot_preds = slot_preds_unmasked[subtokens_mask]
+slot_accuracy = np.mean(slot_labels == slot_preds)
+logging.info(f'Slot Accuracy: {slot_accuracy}')
+f1_scores = get_f1_scores(slot_labels, slot_preds, average_modes=['weighted', 'macro', 'micro'])
+for k, v in f1_scores.items():
+    logging.info(f'{k}: {v}')
+
+logging.info(f'\n {get_classification_report(slot_labels, slot_preds, label_ids=data_desc.slots_label_ids)}')
+
+# intent accuracies
+logging.info('Intent Prediction Results:')
+intent_preds = np.asarray(np.argmax(intent_logits, 1))
+intent_labels = np.asarray(intent_labels)
+intent_accuracy = np.mean(intent_labels == intent_preds)
+logging.info(f'Intent Accuracy: {intent_accuracy}')
+f1_scores = get_f1_scores(intent_labels, intent_preds, average_modes=['weighted', 'macro', 'micro'])
+for k, v in f1_scores.items():
+    logging.info(f'{k}: {v}')
+
+logging.info(f'\n {get_classification_report(intent_labels, intent_preds, label_ids=data_desc.intents_label_ids)}')
+
+# print queries with wrong intent:
+queries = open(f'{data_desc.data_dir}/{args.eval_file_prefix}.tsv', 'r').readlines()[1:]
+intent_dict = open(data_desc.intent_dict_file, 'r').read().splitlines()
+log_misclassified_queries(intent_labels, intent_preds, queries, intent_dict, limit=30)
+
+# print queries with wrong slots:
+slot_dict = open(data_desc.slot_dict_file, 'r').read().splitlines()
+log_misclassified_slots(
+    intent_labels,
+    intent_preds,
+    slot_labels_unmasked,
+    slot_preds_unmasked,
+    subtokens_mask,
+    queries,
+    intent_dict,
+    slot_dict,
+    limit=30,
+)
+
+# analyze confusion matrices
+max_pairs = 20
+logging.info('')
+logging.info(f'*** Most Confused Intents (limit {max_pairs}) ***')
```

Comment:
You may use \n at the beginning of the second logging or have it at then end of previous logging. Instead of havinf empty logging.

Reply:
If I use "'\n" at the beginning then printing of the title will be misaligned (it will start from the start of the string and not after log prefix. I am adding this specifically for better readability. And I did not find nicer way to do it.